### PR TITLE
fix(serenity): make siteId authoritative over brandDomain in Add Market

### PR DESCRIPTION
**Pull request by rainer-friederich**
## 1. Abstract
Makes a supplied `siteId` authoritative over `brandDomain` when creating a Serenity/LLMO market, at both the flat and sub-workspace create-market call sites.

## 2. Reasoning
An Add Market request supplying both `siteId` and `brandDomain` let `brandDomain` win at both create-market call sites, provisioning a Semrush project against the brand's primary domain while linking the mapping row to the caller's selected Site — two conflicting identities on one market row. Confirmed in production for a real customer: two Revlon Canadian markets were provisioned against `revlon.com` instead of the selected `revlon.ca` Site, both published live with zero prompts, rendered indefinitely as "Collecting your data."

Reported in https://github.com/adobe/project-elmo-ui/issues/3031 . That issue also asked for a `project-elmo-ui` frontend change (stop sending `brandDomain` when `siteId` is present), a zero-prompt empty-state UI fix, and manual data repair for the two live Revlon projects. Following a colleague's review comment on the issue narrowing scope to the actual root cause, this PR implements only the API-side fix described here; the frontend change, the empty-state UI work, and the data repair are deferred separately.

## 3. High-level overview of the changes
- `src/support/serenity/handlers/markets.js` (`handleCreateMarket`, flat mode) and `src/controllers/serenity.js` (`createMarket`, sub-workspace mode): a supplied `siteId` is now checked first. When it resolves, its identity is authoritative for the market's Semrush `domain`/`primaryUrl`, regardless of whether a `brandDomain` was also supplied. `brandDomain` is only consulted when no `siteId` was supplied at all. A supplied-but-unresolvable `siteId` still hard-fails with a 400, rather than falling back to `brandDomain`.
- This is deliberately not "reject on mismatch": a brand's primary domain differing from a market's Site domain is the normal, intended Add Market shape (most brands have markets on Sites other than their apex domain) and is never compared or rejected — it is silently superseded whenever a `siteId` resolves.
- Added `resolveMarketIdentity` (`src/support/serenity/site-linkage.js`), the single shared implementation of this precedence rule. The bug existed because the rule was written independently at both call sites and drifted out of sync; both call sites now reference the same function.
- Added `logMarketCreated` (same file), a shared success-telemetry event fired once per create-market success (both modes), reporting brandId, market/geoTargetId, languageCode, siteId, resolved domain, resolved primaryUrl, semrushWorkspaceId, semrushProjectId, `generatePrompts`, and `promptCount` (when applicable). There was previously no structured event recording the outcome of an Add Market call, so it was unverifiable after the fact whether prompt generation was requested and how many prompts a call produced.
- No change to the API's request or response contract — this is a server-side correctness fix plus new logging, not a shape change.

## 4. Required information
- Jira / issue: https://github.com/adobe/project-elmo-ui/issues/3031

## 7. Additional information outside the code
Ran the full `test/it/postgres/serenity.test.js` suite against live Docker Postgres and the Semrush vendor mock containers, including a new case that posts a real `siteId`/`brandDomain` conflict (a Site whose domain is `site2.example.com` alongside a literal `brandDomain: 'example.com'`). Confirmed via the emitted `serenity create-market: market created` log line and a follow-up `GET /serenity/markets/:geo/:lang` that the created project tracked the resolved Site's domain (`site2.example.com`), not the conflicting literal `brandDomain` — proving the fix end-to-end against the real request/response path, not just in isolation.

## 8. Test plan
(a) Locally: exercised the fixed precedence end-to-end via the IT suite described above (real Postgres + Semrush mock, real HTTP request/response), in addition to the unit-level coverage in the diff.

(b) Per-environment verification:
- **dev**: `POST /v2/orgs/:orgId/brands/:brandId/serenity/markets` with both a `siteId` for an onboarded Site and a `brandDomain` that differs from that Site's domain (an existing brand/market combination works, or use a fresh test brand). Follow with `GET /v2/orgs/:orgId/brands/:brandId/serenity/markets/:geoTargetId/:languageCode` and confirm the returned `domain`/`primaryUrl` match the Site, not the supplied `brandDomain`. Check Splunk for the new `serenity create-market: market created` log line to confirm the telemetry fires with the expected fields.

🤖 Generated with [review-kit](https://github.com/adobe/experience-success-skills)

## Change Management

```yaml
cm-assessment: v1
changeType: standard
impact: unnoticeable
risk: minor
changeApprovedBy: ["Dominique Jäggi"]
rationale: "Bug fix changing linkage precedence logic; recoverable via redeploy, not destructive."
```

### CODE REVIEW

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @rainer-friederich,

**Verdict: Request changes** - the core bug fix is correct but the shared extraction is only half-applied in the controller, docs are stale, and telemetry has inconsistencies between the two paths.
**Complexity: HIGH** - medium diff; API surface + multiple converging concerns.
_Changes: makes siteId authoritative over brandDomain in both create-market paths by extracting a shared resolveMarketIdentity function, and adds structured success telemetry (7 files)._

Note: CI checks are still pending - resolve before merge.

## Must fix before merge

1. [Important] Controller subworkspace path does not use resolveMarketIdentity for the brandDomain-only case, undermining the single-source-of-truth intent - `src/controllers/serenity.js:889` (details inline)
2. [Important] docs/serenity.md contains stale text that contradicts the new siteId-authoritative behavior - the "brandDomain OR siteId (LLMO-6405 Phase 2)" section still describes brandDomain as the preferred source and siteId as a fallback alternative; update to reflect siteId-authoritative precedence
3. [Important] Telemetry siteId field is inconsistent between the two call sites (supplied vs linked) - `src/controllers/serenity.js:1017` (details inline)
4. [Important] Telemetry tests assert only the log message string, not the field values that are the point of logMarketCreated - `test/support/serenity/handlers/markets.test.js:534` (details inline)

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: redundant `!brandDomain || !hasText(brandDomain)` guard in flat handler - resolveMarketIdentity never returns empty-string domain, so `!hasText(brandDomain)` alone suffices - `src/support/serenity/handlers/markets.js:408`
- suggestion: add a unit test for `resolveMarketIdentity({ domain: 'example.com', primaryUrl: null }, true, ...)` to pin the null-primaryUrl edge case - `test/support/serenity/site-linkage.test.js`
</details>

File: src/controllers/serenity.js [hunk 1]
```diff
@@ -817,6 +819,13 @@ function SerenityController(context, log, env) {
     }
   };
 
+  /**
+   * @typedef {{
+   *   geoTargetId: number, languageCode: string|null, workspaceId: string,
+   *   promptCount?: number,
+   * }} MarketCreateSuccessBody
+   */
```
Comment (MysticatBot) [inline src/controllers/serenity.js:827]: issue (blocking): The subworkspace path only calls resolveMarketIdentity inside if (suppliedSiteId). When no siteId is supplied, the controller builds effectiveBody with inline derivation (primaryUrl: siteIdentityFromUrlString(requestBody.brandDomain)) instead of going through resolveMarketIdentity. The brandDomain-only path never passes through the shared function.

The flat handler in markets.js delegates unconditionally to resolveMarketIdentity for both cases (lines 157-166). The whole point of extracting the shared function was to keep these two call sites in sync. If resolveMarketIdentity evolves (e.g., normalizing domains, adding validation), the controller inline path will silently diverge.

Additionally, the flat handler resolveMarketIdentity respects a caller-supplied body.primaryUrl when present, while the controller inline path always derives from brandDomain alone, ignoring requestBody.primaryUrl. This is a behavioral asymmetry that undercuts the stated convergence goal.

Fix: Call resolveMarketIdentity(null, false, requestBody.brandDomain, requestBody.primaryUrl) for the no-siteId case and use the returned values for effectiveBody, mirroring how the flat handler does it unconditionally.

File: src/controllers/serenity.js [hunk 2]
```diff
@@ -817,6 +819,13 @@ function SerenityController(context, log, env) {
     }
   };
 
+  /**
+   * @typedef {{
+   *   geoTargetId: number, languageCode: string|null, workspaceId: string,
+   *   promptCount?: number,
```
Comment (MysticatBot) [inline src/controllers/serenity.js:825]: issue (blocking): The telemetry siteId field uses linkedSiteId here (from ensureMarketSite), but the flat handler uses siteIdSupplied ? body.siteId : null (the supplied value). linkedSiteId can be null on best-effort mirror-write failure even when a valid siteId was supplied and resolved. This means the subworkspace path logs siteId: null for a successfully created market that had a valid siteId when the brand_sites mirror write fails, while the flat handler would log the supplied siteId in the same scenario.

Fix: Use the supplied siteId consistently (e.g., suppliedSiteId ?? null) or add a separate linkedSiteId field so ops can distinguish no siteId supplied from siteId supplied but mirror-link failed.

File: test/support/serenity/handlers/markets.test.js [hunk 3]
```diff
@@ -502,8 +503,34 @@ describe('handlers/markets.js — handleCreateMarket', () => {
     }, fakeLog());
 
     expect(result.status).to.equal(201);
-    expect(dataAccess.Site.findById).to.not.have.been.called;
-    expect(transport.createProject.firstCall.args[1].domain).to.equal('adobe.com');
```
Comment (MysticatBot) [inline test/support/serenity/handlers/markets.test.js:506]: issue (blocking): Both telemetry tests (this file and serenity.test.js) assert only the log message string via calledWithMatch regex. This proves the message was logged but not that the fields object contains the correct resolved identity - the entire point of this PR. A regression that logs the wrong domain in the fields object would pass these tests. At minimum, the flat handler test (logs market-created telemetry with the resolved brandDomain/primaryUrl) should assert the domain and primaryUrl fields match the expected resolved values.

**Comment by rainer-friederich** [comment]
Addressed the review in `e7b52076b`.

MysticatBot — Must fix before merge:

- Controller subworkspace path did not use `resolveMarketIdentity` for the brandDomain-only case (`src/controllers/serenity.js:889`) — done. Both branches now route through the one shared function; a caller-supplied `primaryUrl` is still never trusted on this call site, matching the pre-existing security posture (unchanged from before this PR).
- Stale docs describing brandDomain as preferred — done. `docs/serenity.md`'s Add Market section now states the siteId-authoritative precedence explicitly, including the never-falls-back-on-unresolvable-siteId behavior.
- Telemetry siteId field inconsistent between paths (`src/controllers/serenity.js:1017`) — done. The subworkspace path now logs the supplied/resolved siteId (matching the flat handler), not the best-effort mirror-link result, which can be null on a write hiccup even when a valid siteId was supplied.
- Telemetry tests asserted only the log message (`test/support/serenity/handlers/markets.test.js:534`) — done. Both telemetry tests now also assert the resolved field values (domain, primaryUrl, siteId, workspaceId, projectId, generatePrompts/promptCount).

MysticatBot — Non-blocking:

- Redundant `!brandDomain || !hasText(brandDomain)` guard — declined. `hasText` is typed `(str: string): boolean` per this directory's own convention (`src/support/serenity/CLAUDE.md`), so calling it directly on a `string|null` value fails the type-check tier this repo enforces; the `!brandDomain ||` prefix is what narrows the type before the call, not redundant defensive code.
- Add a null-primaryUrl edge case test for `resolveMarketIdentity` — done, in `test/support/serenity/site-linkage.test.js`.

File: src/controllers/serenity.js (hunk 1 above)
Comment (rainer-friederich) [reply src/controllers/serenity.js:827]: Fixed in `e7b52076b` — the no-siteId branch now also routes through `resolveMarketIdentity`. Kept the existing "never trust a caller-supplied primaryUrl on this path" behavior by continuing to pass `undefined` for that argument, so this fix doesn't reopen that separate, pre-existing security posture.

File: src/controllers/serenity.js (hunk 2 above)
Comment (rainer-friederich) [reply src/controllers/serenity.js:825]: Fixed in `e7b52076b` — now logs `suppliedSiteId ?? null` instead of `linkedSiteId`, matching the flat handler's semantics.

File: test/support/serenity/handlers/markets.test.js (hunk 3 above)
Comment (rainer-friederich) [reply test/support/serenity/handlers/markets.test.js:506]: Fixed in `e7b52076b` — both telemetry tests now also assert the resolved field values (domain, primaryUrl, siteId, workspaceId, projectId, generatePrompts/promptCount), not just the log message.

**Review by MysticatBot — APPROVED** [review]
Hey @rainer-friederich,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - all four prior findings are addressed; no new blocking issues.
**Complexity: HIGH** - large diff; API surface.
_Changes: makes siteId authoritative over brandDomain in both create-market paths by extracting a shared resolveMarketIdentity function, and adds structured success telemetry (8 files)._

Note: CI checks are still pending - resolve before merge.
Note: Recommend a human read before merge - this change modifies project documentation (docs/serenity.md) that serves as the operational reference for the Add Market contract. The bot review is a complement to, not a replacement for, a human read here.

## Previously flagged, now resolved

- Controller subworkspace path now routes both branches through resolveMarketIdentity unconditionally
- docs/serenity.md updated to describe siteId-authoritative precedence, hard-fail semantic, and shared implementation
- Telemetry siteId field now uses suppliedSiteId consistently (not linkedSiteId) across both paths
- Telemetry tests assert resolved field values, not just log message strings
- Null-primaryUrl edge case test added for resolveMarketIdentity

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: subworkspace telemetry is inside the `if (projectId)` guard, so a hypothetical 201-without-projectId would silently skip the log; the flat handler logs unconditionally - `src/controllers/serenity.js:1017`
- suggestion: the flat handler's `resolveSiteIdentity` call omits `organizationId` (pre-existing, not introduced by this PR), which means a caller who knows a foreign org's siteId could resolve it on this path; worth a follow-up to thread the org context through - `src/support/serenity/handlers/markets.js:399`
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 3m 21s | <b>Cost:</b> $8.26 | <b>Commit:</b> <code>e7b52076b5a4f79e5affd94137254e09b4d350ee</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by rainer-friederich** [comment]
Addressed the two non-blocking items in `1e8eb5399`, and pushed a separate doc-only correction in `eb7f7d462` unrelated to this review round.

MysticatBot — Non-blocking:

- Subworkspace telemetry inside `if (projectId)` (`src/controllers/serenity.js:1017`) — done. `logMarketCreated` now fires unconditionally on the outer `status === 201`, matching the flat handler's unconditional log; only the DB link (`linkSiteToRow`) stays conditional on a real `projectId`. Added a test asserting the telemetry event fires (with `semrushProjectId: null`) on the same malformed-body case the existing warn-log test already covers.
- `resolveSiteIdentity`'s missing `organizationId` scoping (`src/support/serenity/handlers/markets.js:399`) — declined for this PR. The bot's own note flags it as pre-existing, not introduced here; fixing it means threading org context through a call site this PR only touches for the siteId-vs-brandDomain precedence fix, which is a separate change with its own blast radius. Worth a follow-up.

Unrelated to this review: `eb7f7d462` corrects a factual claim in two code comments (`site-url-propagation.js`, `sites.js`) discovered while investigating data-recovery for the Revlon incident — a Semrush project's `domain` turned out to be immutable in place once published (only `primary_url` and the benchmark move), contradicting an earlier "live-verified" comment. Doc-only; no behavior change.

**Comment by rainer-friederich** [comment]
Could not validate on `dev-branches` (dev CI: `https://spacecat.experiencecloud.live/api/ci/`).

- Confirmed the branch's own deployment landed: GitHub deployment record for commit `1e8eb5399` on environment `dev-branches`, state `success` (`ci / branch-deploy` job green, including its own Post-Deployment Integration Test).
- Attempted a real, mutating exercise of the fixed path: `POST /v2/orgs/{orgId}/brands/{brandId}/serenity/markets` against a genuine dev-onboarded serenity brand (Bark for Schools, sub-workspace mode) with a supplied `siteId` (bark.us) plus a deliberately conflicting `brandDomain` (example.com) — the exact shape of the bug this PR fixes.
- Blocked before the call ever reached the handler: every session token minted for the dev/stage IMS realm (`mysticat login -e dev`, tried with both default and `--product LLMO`) resolved to a low-privilege identity (`is_admin: false`, `is_llmo_administrator: false`, `tenants[0].entitlement.productCode: "ASO"`) distinct from the admin identity this session otherwise runs as on prod IMS — `x-product 'aso' is not authorized for this FACS-governed route`, 403, on a plain `GET .../serenity/markets` too, so this isn't specific to the mutating call. No request reached the fixed code path, and nothing was created — no cleanup needed.
- Logs: not read, since no request reached the deployed code.
- Cleanup: not needed — the blocked call created nothing.

This is an environment/credential gap (a dev/stage-IMS-scoped admin identity), not evidence against the fix. Correctness is otherwise evidenced by the repo's own `test/it/postgres/serenity.test.js` IT suite (Docker Postgres + Semrush vendor mock, 77 passing tests this session, including a case posting this exact siteId-vs-conflicting-brandDomain shape and confirming the created project tracked the resolved site's domain, not the conflicting brandDomain) and by a live production data audit confirming the bug pattern existed and matches this fix's scope.
